File: FFDownload/Utility/Browser/XML/modRSS.vb
Imports System.Xml

Module modRSS

#Region "RSS Feed Conversion"

    Function RDFtoRSS(ByVal xmlDoc As XmlDocument) As XmlDocument

        Dim xml_Doc As New XmlDocument

        xml_Doc = xmlDoc

        Dim xmlList As XmlNodeList = xml_Doc.DocumentElement.ChildNodes

        Dim count As Long
        Dim max As Long = xmlList.Count - 1

        Dim name As String


        For count = max To 0 Step -1
            name = xmlList.Item(count).Name
            Select Case name
                Case "item"
                    'Do Nothing
                Case Else
                    xml_Doc.DocumentElement.RemoveChild(xmlList.Item(count))
            End Select
        Next

        xmlDoc = xml_Doc

        Dim xml As String
        xmlList = xmlDoc.SelectNodes("//item")

        If xmlList.Count = 0 Then
            Return Nothing
        End If

        Dim count2 As Integer

        xml = "<rss>"

        For count2 = 0 To xmlList.Count - 1

            xml += xmlList(count2).OuterXml

        Next

        xml += "</rss>"

        xml_Doc.LoadXml(xml)

        Return xml_Doc

    End Function

    Function ATOMtoRSS(ByVal xmlDoc As XmlDocument) As XmlDocument

        Dim xml_Doc As New XmlDocument

        xml_Doc = xmlDoc

        Dim xmlList As XmlNodeList = xml_Doc.DocumentElement.ChildNodes

        Dim count As Long
        Dim max As Long = xmlList.Count - 1

        Dim name As String

        For count = max To 0 Step -1
            name = xmlList.Item(count).Name
            Select Case name
                Case "entry"
                    'Do Nothing
                Case Else
                    xml_Doc.DocumentElement.RemoveChild(xmlList.Item(count))
            End Select
        Next

        xmlDoc = xml_Doc

        Dim xml As String
        xmlList = xmlDoc.SelectNodes("//entry")

        If xmlList.Count = 0 Then
            Return Nothing
        End If

        Dim count2 As Integer

        xml = "<rss>"

        For count2 = 0 To xmlList.Count - 1

            xml += xmlList(count2).OuterXml

        Next

        xml += "</rss>"

        xml_Doc.LoadXml(xml)


        Return xml_Doc

    End Function

    Function CleanRSS(ByVal xmlDoc As XmlDocument) As XmlDocument

        Dim xml_Doc As New XmlDocument

        xml_Doc = xmlDoc

        Dim node As XmlNode

        Dim xmlList As XmlNodeList = xml_Doc.DocumentElement.ChildNodes

        node = xmlList(0)

        xmlList = node.ChildNodes

        Dim count As Long
        Dim max As Long = xmlList.Count - 1

        Dim name As String

        For count = max To 0 Step -1
            name = xmlList.Item(count).Name.ToLower
            Select Case name
                Case "item"
                    'Do Nothing
                Case Else
                    node.RemoveChild(xmlList.Item(count))
            End Select
        Next


        xmlDoc = xml_Doc

        Dim xml As String
        xmlList = xmlDoc.SelectNodes("//item")

        If xmlList.Count = 0 Then
            Return Nothing
        End If

        Dim count2 As Integer

        xml = "<rss>"

        For count2 = 0 To xmlList.Count - 1

            xml += xmlList(count2).OuterXml

        Next

        xml += "</rss>"

        xml_Doc.LoadXml(xml)


        Return xml_Doc

    End Function

#End Region

    Function CleanFeed(ByVal xmlDoc As XmlDocument) As XmlDocument

        Dim rss As XmlDocument

        rss = xmlDoc

        Dim type As String

        type = rss.DocumentElement.LocalName.ToLower

        Select Case type
            Case "rdf"
                rss = RDFtoRSS(rss)
            Case "feed"
                rss = ATOMtoRSS(rss)
            Case "rss"
                rss = CleanRSS(rss)
            Case Else
                Return Nothing
        End Select

        Return rss

    End Function

End Module

File: FFDownload/Utility/Browser/XML/modRSS.cs
using System.Xml;

namespace HtmlGrabber
{

    static class modRSS
    {

        #region RSS Feed Conversion

        public static XmlDocument RDFtoRSS(XmlDocument xmlDoc)
        {

            var xml_Doc = new XmlDocument();

            xml_Doc = xmlDoc;

            var xmlList = xml_Doc.DocumentElement.ChildNodes;

            long count;
            long max = xmlList.Count - 1;

            string name;


            for (count = max; count >= 0L; count += -1)
            {
                name = xmlList.Item((int)count).Name;
                switch (name ?? "")
                {
                    // Do Nothing
                    case "item":
                        {
                            break;
                        }

                    default:
                        {
                            xml_Doc.DocumentElement.RemoveChild(xmlList.Item((int)count));
                            break;
                        }
                }
            }

            xmlDoc = xml_Doc;

            string xml;
            xmlList = xmlDoc.SelectNodes("//item");

            if (xmlList.Count == 0)
            {
                return null;
            }

            int count2;

            xml = "<rss>";

            var loopTo = xmlList.Count - 1;
            for (count2 = 0; count2 <= loopTo; count2++)


                xml += xmlList[count2].OuterXml;

            xml += "</rss>";

            xml_Doc.LoadXml(xml);

            return xml_Doc;

        }

        public static XmlDocument ATOMtoRSS(XmlDocument xmlDoc)
        {

            var xml_Doc = new XmlDocument();

            xml_Doc = xmlDoc;

            var xmlList = xml_Doc.DocumentElement.ChildNodes;

            long count;
            long max = xmlList.Count - 1;

            string name;

            for (count = max; count >= 0L; count += -1)
            {
                name = xmlList.Item((int)count).Name;
                switch (name ?? "")
                {
                    // Do Nothing
                    case "entry":
                        {
                            break;
                        }

                    default:
                        {
                            xml_Doc.DocumentElement.RemoveChild(xmlList.Item((int)count));
                            break;
                        }
                }
            }

            xmlDoc = xml_Doc;

            string xml;
            xmlList = xmlDoc.SelectNodes("//entry");

            if (xmlList.Count == 0)
            {
                return null;
            }

            int count2;

            xml = "<rss>";

            var loopTo = xmlList.Count - 1;
            for (count2 = 0; count2 <= loopTo; count2++)


                xml += xmlList[count2].OuterXml;

            xml += "</rss>";

            xml_Doc.LoadXml(xml);


            return xml_Doc;

        }

        public static XmlDocument CleanRSS(XmlDocument xmlDoc)
        {

            var xml_Doc = new XmlDocument();

            xml_Doc = xmlDoc;

            XmlNode node;

            var xmlList = xml_Doc.DocumentElement.ChildNodes;

            node = xmlList[0];

            xmlList = node.ChildNodes;

            long count;
            long max = xmlList.Count - 1;

            string name;

            for (count = max; count >= 0L; count += -1)
            {
                name = xmlList.Item((int)count).Name.ToLower();
                switch (name ?? "")
                {
                    // Do Nothing
                    case "item":
                        {
                            break;
                        }

                    default:
                        {
                            node.RemoveChild(xmlList.Item((int)count));
                            break;
                        }
                }
            }


            xmlDoc = xml_Doc;

            string xml;
            xmlList = xmlDoc.SelectNodes("//item");

            if (xmlList.Count == 0)
            {
                return null;
            }

            int count2;

            xml = "<rss>";

            var loopTo = xmlList.Count - 1;
            for (count2 = 0; count2 <= loopTo; count2++)


                xml += xmlList[count2].OuterXml;

            xml += "</rss>";

            xml_Doc.LoadXml(xml);


            return xml_Doc;

        }

        #endregion

        public static XmlDocument CleanFeed(XmlDocument xmlDoc)
        {

            XmlDocument rss;

            rss = xmlDoc;

            string type;

            type = rss.DocumentElement.LocalName.ToLower();

            switch (type ?? "")
            {
                case "rdf":
                    {
                        rss = RDFtoRSS(rss);
                        break;
                    }
                case "feed":
                    {
                        rss = ATOMtoRSS(rss);
                        break;
                    }
                case "rss":
                    {
                        rss = CleanRSS(rss);
                        break;
                    }

                default:
                    {
                        return null;
                    }
            }

            return rss;

        }

    }
}